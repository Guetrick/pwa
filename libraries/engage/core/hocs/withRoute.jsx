import React from 'react';
import { RouteContext } from '@shopgate/pwa-common/context';

/**
 * Gets the context props for the wrapped component.
 * @param {Object} context The context.
 * @param {string|null} prop The optional prop name.
 * @returns {Object}
 */
const getInjectedProps = (context, prop) => {
  if (!prop) {
    return context;
  }

  return {
    [prop]: { ...context },
  };
};

/**
 * Injects the route properties into the desired component.
 * @param {Function} WrappedComponent The react component to wrap.
 * @param {Object} [options={}] Options for the HOC.
 * @param {string} [options.prop] An optional prop name to inject the route properties.
 * @returns {JSX}
 */
export function withRoute(WrappedComponent, options = {}) {
  /**
   * The actual HOC.
   * @param {Object} props The component props.
   * @returns {JSX}
   */
  const WithRoute = props => (
    <RouteContext.Consumer>
      {routeContext => (
        <WrappedComponent {...props} {...getInjectedProps(routeContext, options.prop)} />
      )}
    </RouteContext.Consumer>
  );

  const displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component';
  WithRoute.displayName = `WithRoute(${displayName})`;

  return WithRoute;
}
