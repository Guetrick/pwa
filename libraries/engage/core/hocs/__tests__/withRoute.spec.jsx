import React from 'react';
import { mount } from 'enzyme';
import { withRoute } from '../withRoute';

const mockRoute = {
  id: '32d55',
  pathname: '/',
  pattern: '/',
  state: {},
};

const mockContext = jest.fn().mockReturnValue(mockRoute);

jest.mock('@shopgate/pwa-common/context', () => ({
  RouteContext: {
    Consumer: ({ children }) => children(mockContext()),
  },
}));

// eslint-disable-next-line require-jsdoc
const MockComponent = () => null;

describe('engage > core > hocs > withRoute', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should inject the context properties into the component', () => {
    const ComposedComponent = withRoute(MockComponent);
    const wrapper = mount(<ComposedComponent someProp />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).props()).toEqual({
      someProp: true,
      ...mockRoute,
    });
  });

  it('should inject a single property with the context into the component', () => {
    const ComposedComponent = withRoute(MockComponent, { prop: 'route' });
    const wrapper = mount(<ComposedComponent someProp />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).props()).toEqual({
      someProp: true,
      route: {
        ...mockRoute,
      },
    });
  });
});
