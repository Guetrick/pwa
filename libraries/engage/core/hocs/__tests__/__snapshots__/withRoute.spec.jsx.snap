// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > core > hocs > withRoute should inject a single property with the context into the component 1`] = `
<WithRoute(MockComponent)
  someProp={true}
>
  <Consumer>
    <MockComponent
      route={
        Object {
          "id": "32d55",
          "pathname": "/",
          "pattern": "/",
          "state": Object {},
        }
      }
      someProp={true}
    />
  </Consumer>
</WithRoute(MockComponent)>
`;

exports[`engage > core > hocs > withRoute should inject the context properties into the component 1`] = `
<WithRoute(MockComponent)
  someProp={true}
>
  <Consumer>
    <MockComponent
      id="32d55"
      pathname="/"
      pattern="/"
      someProp={true}
      state={Object {}}
    />
  </Consumer>
</WithRoute(MockComponent)>
`;
